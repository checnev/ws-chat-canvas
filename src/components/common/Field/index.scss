.input {
  width: 100%;
  border: 1px solid rgba(0, 0, 0, 0.25);
  padding: 10px 12px;
  margin-bottom: 3px;

  &_invalid {
    border-color: rgb(233, 86, 86);
  }

  &__label {
    display: block;

    &_hasLabel {
      padding-top: 10px;
    }
  }

  &__helpMessage {
    display: block;
    color: rgb(233, 86, 86);
  }

  &__fieldWrapper {
    position: relative;
  }

  &__labelText {
    position: absolute;
    top: -10px;
    left: 12px;
    background: rgb(255, 255, 255);
    padding: 2px 5px;
    color: rgb(153, 153, 153);
    font-size: .8em;
  }
}
