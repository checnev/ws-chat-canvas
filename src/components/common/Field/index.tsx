import React from 'react';
import clsx from 'clsx';
import cn from 'utils/cn';
import './index.scss';

interface FieldProps extends React.InputHTMLAttributes<HTMLInputElement> {
  label?: string;
  invalid?: boolean;
  helpMessage?: string;
}

class Field extends React.Component<FieldProps> {
  render() {
    const { 
      type = 'text',
      className,
      invalid,
      label,
      helpMessage,
      ...props
    } = this.props;
    const input = cn('input');
    return (
      <label className={input('label', { hasLabel: Boolean(label) })}>
        <div className={input('fieldWrapper')}>
          {label && <span className={input('labelText')}>{label}</span>}
          <input
            type={type}
            className={
              clsx(className, input({ type, invalid }))
            }
            {...props}
          />
        </div>
        {invalid && helpMessage && <small className={input('helpMessage')}>{helpMessage}</small>}
      </label>
    );
  }
}

export default Field;
