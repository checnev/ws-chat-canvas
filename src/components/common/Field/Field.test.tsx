import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import pretty from 'pretty';
import Field from 'components/common/Field';

let container: any = null;

beforeEach(() => {
  container = document.createElement('div');
  document.body.append(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe('Field component', () => {
  it('should render Field', () => {
    act(() => {
      render(<Field type="text" label="Field Label" />, container);
    });

    expect(
      pretty(container.innerHTML)
    ).toMatchSnapshot();
  });

  it('should render Field with helpMessage', () => {
    act(() => {
      render(<Field type="text" label="Field Label" helpMessage="invalid value" invalid />, container);
    });

    expect(
      pretty(container.innerHTML)
    ).toMatchSnapshot();
  });
});
