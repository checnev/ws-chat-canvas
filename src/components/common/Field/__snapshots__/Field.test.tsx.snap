// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field component should render Field 1`] = `
"<label class=\\"input__label input__label_hasLabel\\">
  <div class=\\"input__fieldWrapper\\"><span class=\\"input__labelText\\">Field Label</span><input type=\\"text\\" class=\\"input input_type_text\\"></div>
</label>"
`;

exports[`Field component should render Field with helpMessage 1`] = `
"<label class=\\"input__label input__label_hasLabel\\">
  <div class=\\"input__fieldWrapper\\"><span class=\\"input__labelText\\">Field Label</span><input type=\\"text\\" class=\\"input input_type_text input_invalid\\"></div><small class=\\"input__helpMessage\\">invalid value</small>
</label>"
`;
